File: 105/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        unordered_map<int, int> inorderTable;
        
        for (int i=0; i<inorder.size(); i++) {
            inorderTable[inorder[i]] = i;
        }
        return solve(preorder, 0, inorder, 0, inorder.size(), inorderTable);
    }
    
private:
    TreeNode* solve(vector<int>& preorder, int pivot, vector<int>& inorder, int inI, int inJ, unordered_map<int, int>& inorderTable) {
        if (pivot >= preorder.size() || inI >= inJ) return NULL;
        int midElement = preorder[pivot];
        
        int found = inorderTable[midElement];
        int distance = found - inI;
            
        TreeNode *treeNode = new TreeNode(midElement);
        treeNode->left = solve(preorder, pivot+1, inorder, inI, inI + distance, inorderTable);
        treeNode->right = solve(preorder, pivot+distance+1, inorder, inI + distance + 1, inJ, inorderTable);
        return treeNode;
    }
};

File: 105/solution.go
// In Go, we don't need to do index passing like C, it can directly pass subarray to next recursion
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func buildTree(preorder []int, inorder []int) *TreeNode {
	inorderMap := make(map[int]int)

	for i := 0; i < len(inorder); i++ {
		inorderMap[inorder[i]] = i
	}
	return solve(preorder, 0, inorder, 0, len(inorder), inorderMap)
}

func solve(preorder []int, pivot int, inorder []int, inI int, inJ int, inorderMap map[int]int) *TreeNode {
	if pivot >= len(preorder) || inI >= inJ {
		return nil
	}

	midElement := preorder[pivot]

	found := inorderMap[midElement]
	distance := found - inI

	node := TreeNode{Val: midElement}
	node.Left = solve(preorder, pivot+1, inorder, inI, inI+distance, inorderMap)
	node.Right = solve(preorder, pivot+distance+1, inorder, inI+distance+1, inJ, inorderMap)

	return &node
}